### DISCUSSION TITLE
Deprecating some Company backends

### DISCUSSION THREAD

**Discussion by minad** [company-mode/company-mode#1297] [Ideas]
Hi Dmitry,

today [Cape](http://elpa.gnu.org/packages/cape.html) was added to GNU ELPA. I wonder if it makes sense to deprecate or remove some of the Company backends and instead make use of the Capf backends provided by Cape. Of course this only applies to the simple Capf backends and does not affect any of the asynchronous backends (we can still pick up the async discussion again). I only think about Capfs/backends like `company-keywords` or `cape-keywords`. As far as I understand there is a blocker however. Currently one can group `(company-dabbrev company-keywords)` for example. If these become Capfs (as implemented in Cape), then you would lose this grouping except if Company acquires the ability to group specific Capfs. For illustration, these Capfs could be specified in a hypothetical syntax like this: `(company-capf:cape-dabbrev, company-capf:cape-keywords)`.

Daniel

**Comment by dgutov** [comment]
Hi Daniel,

I'm not sure about the cost/benefit ration of such a move yet. Since these backends don't change as much, the complexity saving from reusing code aren't going to be a big win either.

And Company's `company-backends` format has a significant transparency advantage over all of the other options I'm aware of. When you print the value of `company-backend` it's either (almost certainly) a named function, or a "simple" list of functions. That's easier to recognize and reason about that a bunch of anonymous functions values (possibly even byte-compiled ones) that an average wrapper would provide.

But on a related note, here's an issue that could benefit from a unification: https://github.com/company-mode/company-mode/pull/1215. Does Cape's counterpart to `company-dabbrev-code` both support completion styles (which would make sense) and provide a fast enough completion source for symbols across all buffers? I imagine it needs a robust caching strategy.

**Reply by minad** [reply]
> But on a related note, here's an issue that could benefit from a unification: #1215. Does Cape's counterpart to company-dabbrev-code both support completion styles (which would make sense) and provide a fast enough completion source for symbols across all buffers? I imagine it needs a robust caching strategy.

`cape-dabbrev` reuses dabbrev.el to produce all the candidates. Besides that it is a simple Capf, which of course respects completion styles for filtering. However since it uses dabbrev.el to produce the candidates, the candidates which are found must match the initial prefix, so one could say that at that point, the completion style is not respected, since completion styles don't necessary require a prefix match.

I've seen that you have your own dabbrev implementation in `company-dabbrev` which deviates from dabbrev.el. There is also a dabbrev implementation in hippie.el btw. It seems to me that there are far too many different dabbrev implementations around, therefore I tried to reuse dabbrev.el in `cape-dabbrev`. ;)

But anyways if you see some potential for unification we could also create a `dabbrev-capf` package which could be used by both Company and Corfu. Or you just point users to `cape-dabbrev`? However there is still the aforementioned grouping issue. One could also try to overhaul dabbrev.el to work better as foundation for completion backends. Alternatively we could create a better dabbrev2.el package, which supersedes dabbrev.el with an improved and more flexible API. Did you mean something like that?

**Reply by dgutov** [reply]
I didn't have any particular resolution in mind, but if the built-in `dabbrev` package (or an alternative `dabbrev2` one in ELPA) were to get an easy programmatic API that could support completion styles too, it would go a long way toward code reuse.

If all that is left in `company-dabbrev` and `company-dabbrev-code` are simple wrappers and a bunch of user options, I suppose it would be as good outcome as any?

**Reply by minad** [reply]
Can you clarify what you mean by "supports completion styles"? Do you mean that the dabbrev candidate finding in all the buffers should go via completion styles? It should find all the candidates which match the initial input, however not filtered by prefix but by completion style? This is doable but maybe slow.

**Reply by dgutov** [reply]
Yes, it would probably return a flat list of strings (all symbols or words in related buffers). Probably searched based on some regexp, so that the notion of a "word" is customizable as well. Just as the notion of "related buffers".

> This is doable but maybe slow.

Hence the need for some caching strategy. I tried to describe my (very rough) vision in the second-to-last paragraph in the comment here: https://github.com/company-mode/company-mode/pull/1215#issuecomment-933094051

**Reply by minad** [reply]
Sounds good. One could extract all words, cache them somewhere and update if the buffer ticks change. Then the many resulting candidates can just be returned as completion table/candidates and the filtering is done completely by completion styles. Let's create such a package and you deprecate company-dabbrev and I remove cape-dabbrev.

We can add the company and capf backends as part of this package, such that we don't create any kind of dependency besides the Emacs core. It only requires the Company backend to be written in a style which does not depend on Company as we discussed before in the async capf discussion. And well, the Capf provided by that package will of course also depends only on builtins. What do you think?

**Reply by minad** [reply]
On top we should probably also implement the existing dabbrev API, dabbrev-expand/dabbrev-completion in order to make dabbrev.el fully obsolete.

**Reply by dgutov** [reply]
> We can add the company and capf backends as part of this package, such that we don't create any kind of dependency besides the Emacs core.

I think at step one we should keep the Company backend inside Company, while moving the implementation "outside". If only to keep backward compatibility.

> It only requires the Company backend to be written in a style which does not depend on Company as we discussed before in the async capf discussion.

I'm up for that direction, but we still haven't solved a couple of "quality of life" issues that would accompany such a move.

And is there a particular reason for us to hurry about that? Do you have any serious plans to migrate to Company-style backends? `dabbrev2` can include a Capf source, of course, and most of the frontends support that.

> On top we should probably also implement the existing dabbrev API, dabbrev-expand/dabbrev-completion in order to make dabbrev.el fully obsolete.

That idea strongly suggests putting the new code into `dabbrev.el` in the core, and then making it a "core ELPA" package... but I remembered now that it might not be your preferred development vehicle.

What would be the long-term plan with "obsoleting dabbrev.el", then? Developing `dabbrev2` for a number of years externally and then proposing to replace `dabbrev.el` in the core wholesale, squashing all the development history?

**Reply by minad** [reply]
The idea to keep the company backend in dabbrev2 was just to avoid adding a dependency to company-mode. You could also detect the presence dynamically and throw an error if dabbrev2 is not present. Besides that there is no reason, except putting the two apis side by side as equals. This would help judging how to go forward in the long term. If once again the company backend would live in company the argument that the company backends are actually frontend agnostic would be lost.

> I think at step one we should keep the Company backend inside Company, while moving the implementation "outside". If only to keep backward compatibility.

Maybe this just means to only keep an alias inside Company, or a small wrapper which adds the interactive support?

> That idea strongly suggests putting the new code into dabbrev.el in the core, and then making it a "core ELPA" package... but I remembered now that it might not be your preferred development vehicle.

Yes I doubt that such an approach is effective and successful. We should develop some dabbrev2 or xabbrev outside and then replace dabbrev in a single act. In my opinion there is nothing in dabbrev.el we can salvage, even if I tried to do that in cape-dabbrev. Dabbrev.el is still not good code, and I had to jump through hoops to make it work. You can judge better of course regarding the best approach or you may discuss it with Stefan for example. In the worst case one ends up with an obsolete dabbrev.el in core and xabbrev.el outside. This is not much different from the situation now, since company-dabbrev.el already contains its own implementation. Furthermore hippie contains another dabbrev implementation. Regarding maintainance status, code duplication and redundancy the current situation doesn't look good anyways. To make it clear - 1. I don't have interest in long discussions to get started and 2. I lack the time and the patience to go through a step by step review for each small change. Note that things like nativecomp or pgtk were also added at once and these things are more complex by dimensions.

**Reply by minad** [reply]
To add one more argument - since the plan is to change the candidate generation method with caching and completion style support, the implementation will also look differently from dabbrev.el. This doesn't make dabbrev.el a good basis to start with. Using dabbrev.el as basis would be the only reason which would justify an in-tree approach.

Maybe the different implementation strategy and the Emacs backward compatibility story is also justification enough to just keep both dabbrev and dabbrev2 around forever? Would it hurt? I don't think it would technically. The current Emacs development style seems unsustainable in the long term. Emacs should better be debundled and decomposed in ELPA packages which are put together for release and distribution. This would distribute the development load and maybe some of the outdated packages in tree would also see more improvements.

**Reply by dgutov** [reply]
> Besides that there is no reason, except putting the two apis side by side as equals. This would help judging how to go forward in the long term. If once again the company backend would live in company the argument that the company backends are actually frontend agnostic would be lost.

I'm just trying to define smaller steps here. Moving the implementation is the hard part technically (as far as the dabbrev package goes), whereas the backend definition is something much thinner.

But there are things that can be changed about the backend API as well (recall the `company-files` and boundaries/fields sub-thread), which would be better to finalize before announcing the push for the public API alternative.

> Maybe this just means to only keep an alias inside Company, or a small wrapper which adds the interactive support?

That can work too. And it would amount to something I also meant to suggest: perhaps giving the new format a new name (it's not a "Company backend" if it's not tied to Company, is it?), and then it can easily have a different set of features. As long as it still stays close, so that the functions are compatible enough.

> Note that things like nativecomp or pgtk were also added at once and these things are more complex by dimensions.

True. I mean, losing development history is not a grave problem, but it's something I try to avoid when feasible.

> I lack the time and the patience to go through a step by step review for each small change.

FWIW, the `dabbrev` package in the core lacks any active development/maintenance, so it should be easy enough to get you appointed to it. And committers don't have to go through a review process.

**Reply by dgutov** [reply]
> This doesn't make dabbrev.el a good basis to start with. Using dabbrev.el as basis would be the only reason which would justify an in-tree approach.

No, I didn't mean that the current implementation is useful as a base. But a smooth, backward-compatible transition could bring the new features (better performance? I don't know what else we actually want in features) to a wider audience sooner. Or maybe I just like the image of having the older code deleted on the master branch sooner.

Anyway, making this development in the core is certainly not a hard requirement. Just making observations.

> Emacs should better be debundled and decomposed in ELPA packages which are put together for release and distribution. This would distribute the development load and maybe some of the outdated packages in tree would also see more improvements.

While I generally agree, I don't believe we'll see this for non-optional features (like ones with default key bindings) anytime soon. Before Emacs gets a much bigger team anyway.

> Maybe the different implementation strategy and the Emacs backward compatibility story is also justification enough to just keep both dabbrev and dabbrev2 around forever? Would it hurt? I don't think it would technically.

Perhaps it would only hurt me, aesthetically. ;-(

**Reply by minad** [reply]
> FWIW, the `dabbrev` package in the core lacks any active development/maintenance, so it should be easy enough to get you appointed to it. And committers don't have to go through a review process.

I am not really interested in maintaining dabbrev. As I wrote above, I think dabbrev is code which should be abandoned instead of trying to salvage it. In particular if we change the entire mechanism with using a cache of all the words etc. But besides that I agree with your points. I would also like to see code being deleted earlier etc. I won't have time for the next few weeks to hack on anything significant but I will come back to this discussion after that.
